### ISSUE TITLE
Problem with "deep" multistage builds: `layer does not exist` error

### ISSUE THREAD

**Issue by jcberthon** [moby/moby#37340]
**Description**

We are having a problem with the multistage builds where on first run we have the `layer does not exist` error, but if we run the build again it works. So on a clean system (with no cache) it fails, but not on a dirty system with cache.

I have first reported this problem in issue #33974 (see comment https://github.com/moby/moby/issues/33974#issuecomment-399445187). But it looks to be distinct, so I took the advice of @thaJeztah and created a new issue.

**Steps to reproduce the issue:**
I have setup a GitHub repo with all the code to reproduce the problem: https://github.com/jcberthon/multistage-fails

In the readme, there are the instructions (step-by-step to reproduce the bug). You can find also all Dockerfiles. The Dockerfile for which the multistage build fails is under the `level3` directory.

Of course you can build every images locally (it fails at level3), or you can directly try to build level3 as I have pushed to Docker Hub the intermediate images.

**Describe the results you received:**

The first time one runs the above example project, he gets the following result:
When building `level3` it fails with the error `layer does not exist` (see the GitHub project above for more details).

**Describe the results you expected:**

When running it a second time, it works as I expect it. Albeit, it should run like that on the first try already.

**Additional information you deem important (e.g. issue happens only occasionally):**

The bug is triggered always on a clean system, never when running a 2nd time. So `docker system prune -a` helps to trigger the bug again.

**Output of `docker version`:**

```
Client:
 Version:      18.05.0-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   f150324
 Built:        Wed May  9 22:16:04 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.05.0-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   f150324
  Built:        Wed May  9 22:19:55 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 18.05.0-ce
Storage Driver: btrfs
 Build Version: Btrfs v4.15.1
 Library Version: 102
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.16.13-300.fc28.x86_64
Operating System: Fedora 28 (Workstation Edition)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.852GiB
Name: spcbo-jcb-fedora.vcs.de
ID: 7ENM:NU5S:7XAB:K2MJ:CVKB:6BUO:HIMM:NW3M:QZ35:KNL6:VCOT:KBRM
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

I have tested this behaviour on 3 different platforms:
* VM VBox: Fedora 28 (Kernel 4.16.13-300.fc28.x86_64) with Docker CE 18.05.0-ce - btrfs
* Bare metal: CentOS 7.5 (Kernel 3.10.0-862.3.2.el7.x86_64) with Docker CE 18.03.1-ce - devicemapper
* Bare metal: Ubuntu 16.04.4 (Kernel 4.13.0-45-generic) with Docker CE 18.03.1-ce - btrfs

**Comment by ygnmhdtt** [comment]
FYI: 

> Steps to reproduce the issue:
> I have setup a GitHub repo with all the code to reproduce the problem: 
> https://github.com/jcberthon/multistage-fails

I tried just like this, it worked correctly.
My environment:

```sh
$ docker version
Client:
 Version:      18.06.0-ce-rc2
 API version:  1.38
 Go version:   go1.10.3
 Git commit:   a4637b4
 Built:        Sat Jul  7 00:02:43 2018
 OS/Arch:      linux/amd64
 Experimental: false

Server:
 Engine:
  Version:      18.06.0-ce-rc2
  API version:  1.38 (minimum version 1.12)
  Go version:   go1.10.3
  Git commit:   a4637b4
  Built:        Sat Jul  7 00:00:47 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Comment by matthax** [comment]
Experiencing a similar issue. Three stage build with multiple copy commands, the issue occurs on the last copy for me.

### docker info
```Containers: 4
 Running: 3
 Paused: 0
 Stopped: 1
Images: 259
Server Version: 18.03.1-ce
Storage Driver: overlay
 Backing Filesystem: xfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.11.6.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 31.26GiB
Name: baby-server.test.net
ID: QVYK:BVXM:XWJW:HBGS:F2UH:2RAL:3BYN:CUBI:CA6H:SVAE:XQKL:LNVO
Docker Root Dir: /app/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

Relevant dockerfile lines:
```
COPY certs /app/certs
COPY main /app/main
COPY static /app/static
COPY wait-for-it.sh commentary.sh gunicorn.conf logging.conf nginx.conf manage.py /app/
```

If I restart my build, it will work, but the first time for an image it will fail.

**Comment by thaJeztah** [comment]
I tried reproducing the issue on Docker 18.06 (Docker for Mac, using overlay2), but so far haven't been able to get it to fail. I'll try setting up other machines with other distros to see if I can reproduce the problem.

**Comment by jcberthon** [comment]
I have cloned again my repo on another machine. It's a bare metal machine running Ubuntu 18.04.1 LTS (Kernel 4.15.0-30-generic) with Docker CE 18.06.

The build fails in exactly the same manner as described in the README on my GitHub repo.

```console
$ docker version
Client:
 Version:           18.06.0-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        0ffa825
 Built:             Wed Jul 18 19:09:54 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.0-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       0ffa825
  Built:            Wed Jul 18 19:07:56 2018
  OS/Arch:          linux/amd64
  Experimental:     false
$ docker system info
Containers: 4
 Running: 4
 Paused: 0
 Stopped: 0
Images: 21
Server Version: 18.06.0-ce
Storage Driver: btrfs
 Build Version: Btrfs v4.15.1
 Library Version: 102
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: d64c661f1d51c48782c9cec8fda7604785f93587
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-30-generic
Operating System: Ubuntu 18.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 23.53GiB
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Comment by jcberthon** [comment]
However, it works on my Raspberry Pi which is running also Docker CE 18.06 but uses overlay2 as storage backend.

So it seems that it is working only for overlay2, but not for BTRFS or Device Mapper.

**Comment by jcberthon** [comment]
Hi @thaJeztah 

Any news on this issue? What's the way forward?

**Comment by thaJeztah** [comment]
Let me ask @kolyshkin if he has an idea; he's been looking at btrfs and devicemapper improvements recently, so may have ideas as to what could cause this.

Curious; is there a specific reason for using BTRFS in your situation?

**Comment by jcberthon** [comment]
At work, we are using devicemapper-only.

At home I decided several years ago to play with fire and I've setup a mini-server with Ubuntu 12.04 (now 18.04) and btrfs for file system. I'm using it for backup and storing photos (note it is not my only backup, hence I can dare playing with fire). Why btrfs? Well it is slower, and 6 years ago there was a lot of file system full when it was only requiring some balancing, so I must have been masochistic somewhat... But since then it is still sluggish (my IO wait and queues are quite high for such a small low activity home server!!) but it has become reliable. One advantage on using btrfs (and one of the reasons I chose it in the first place) is data checksum. On my NAS - which is using ext4 - it has happen that some photos got corrupted (bit rot) and it is time consuming to find an older version which is not affected. On my mini home server, if it did happen (and I had 2 bad very old HDDs with bad sectors which I've replace since then) I never saw it in my data.

6 years ago I've considered between FreeBSD+ZFS or Ubuntu+BTRFS. The latter won for 2 reasons: my platform was AMD and bhyve was not working on it back then, so I needed KVM; and I'm much more familiar with Linux than BSD so it would have been easier to setup. So I went with Ubuntu+BTRFS+KVM (using LVM for the KVM guests, not btrfs+qcow). But my underpower mini-server was quickly not able to run all services within VMs and that's why I started using containers instead, first LXD and then Docker. And when I started with Docker, I was not thinking of what should be my backend storage but rather how do I launch a hello world ;-)

In the end, I'm using btrfs for historical reason and it is the first time I see a bug on this backend with Docker. However, note that I'm not using this deep level multi-stage build at home, I only used that platform to see if the behaviour was reproducible there.

**Comment by jcberthon** [comment]
Hi @kolyshkin 

How could we move this issue forward? I'm a bit constrained in the coming 2 weeks, but I could spare some time helping after.

**Comment by kolyshkin** [comment]
Reproduced locally with 18.06.1-ce + devmapper.

**Comment by kolyshkin** [comment]
First run:
> Step 2/6 : FROM jcberthon/level1:latest as level1
>  ---> 3554ce43a483
> Step 3/6 : FROM jcberthon/level2:latest
>  ---> eee894dc77ac
> Step 4/6 : COPY --from=baseline /project/test/bug.txt /project/test/bug.txt
>  ---> 7dce12f6cbad
> Step 5/6 : COPY --from=level1 /project/test/hello.txt /project/test/hello.txt
> failed to export image: failed to create image: failed to get layer sha256:3f86bfb961b706073627854ade1a512b383d9ff97482ec4080a403f414a65eba: layer does not exist

Second run:
> Step 2/6 : FROM jcberthon/level1:latest as level1
>  ---> 3554ce43a483
> Step 3/6 : FROM jcberthon/level2:latest
>  ---> eee894dc77ac
> Step 4/6 : COPY --from=baseline /project/test/bug.txt /project/test/bug.txt
>  ---> Using cache
>  ---> 7dce12f6cbad
> Step 5/6 : COPY --from=level1 /project/test/hello.txt /project/test/hello.txt
>  ---> 1bd7ac655938
> Step 6/6 : ENTRYPOINT [ "bash", "-c", "cat /project/test/hello.txt | /usr/games/cowsay" ]
>  ---> Running in 117a9e1d50ba
... and it works

**Comment by kolyshkin** [comment]
Reproduced on dockerd built from latest moby/moby + devicemapper.

Interesting, when using buildkit it is not reproducible (`export DOCKER_BUILDKIT=1`).

**Comment by kolyshkin** [comment]
Reproduced on overlay2 (otherwise same setup as prev comment):

> Step 5/6 : COPY --from=level1 /project/test/hello.txt /project/test/hello.txt
> failed to export image: failed to create image: failed to get layer sha256:b8193d60c7afa0d348179db7f260ff89e7fec68c2012727e8ae26b8f28cbd065: layer does not exist
> [root@kir-ce71-2 multistage-fails]# docker info | grep ora
> Storage Driver: overlay2

This means the problem is not in graph driver, but in the general mechanism of referring to images/layers etc.

**Comment by kolyshkin** [comment]
docker-ce-17.05 works, so it's a regression.

**Comment by kolyshkin** [comment]
The regression first appeared in 17.07.0-ce (which fails). Previous version from the centos repo, 17.06.2-ce, works.

**Comment by kolyshkin** [comment]
Looking into the diff between 17.06 and 17.07, this PR looks slightly suspicious: https://github.com/moby/moby/pull/33755

**Comment by kolyshkin** [comment]
Nope, this is something else, I will try to bisect. @jcberthon if you have time, feel free to do that, too.

**Comment by kolyshkin** [comment]
git bisecting in progress...

**Comment by kolyshkin** [comment]
bisected down to https://github.com/moby/moby/pull/33454, rechecking to be sure // CC @dnephin

**Comment by kolyshkin** [comment]
Related: https://github.com/moby/moby/issues/33974. It says #33454 broke it and then #34063 fixes it. Hmm...

**Comment by kolyshkin** [comment]
OK, #33454 breaks it and then #34063 **does not** fix it for me.

**Comment by kolyshkin** [comment]
Another PR, https://github.com/moby/moby/pull/35579, does not fix it either.

@jcberthon and soo I did the full circle.

Apparently what is causing the issue is COPY statement that does not change the container filesystem in any way (as described in https://github.com/moby/moby/issues/33974#issuecomment-353990074).

In this very case, it is this line from your `level3/Dockerfile`:
 COPY --from=baseline /project/test/bug.txt /project/test/bug.txt

It does nothing because `level3` image is derived from `level2` which is derived from `baseline` and so it already contains `/project/test/bug.txt`, so COPY does not change the state of the container fs.

Now, if you comment out that statement, everything works, and the image is the same as it should be.

Still, this looks like a bug to me. @tonistiigi do you have any input on this? Here's the simplest reproducer:

```bash
#!/bin/bash
set -eux

touch 1 2

cat << EOF > Dockerfile
FROM scratch AS base
COPY 1 /dst/1

FROM base
# useless copy works...
COPY --from=base /dst/1 /dst/1
# ... and the statement after it fails
COPY 2 /dst/2
EOF

docker build --no-cache .
```

It fails on every docker release > 17.06:
```
+ touch 1 2
+ cat
+ docker build --no-cache .
Sending build context to Docker daemon  4.096kB
Step 1/5 : FROM scratch AS base
 ---> 
Step 2/5 : COPY 1 /dst/1
 ---> d1dec26e649a
Step 3/5 : FROM base
 ---> d1dec26e649a
Step 4/5 : COPY --from=base /dst/1 /dst/1
 ---> e429e9812ce2
Step 5/5 : COPY 2 /dst/2
failed to export image: failed to create image: failed to get layer sha256:d190fa3fee94e4c0ddd6ab1cebb9bef9e6bc09805b10b4e08870eb62d377d131: layer does not exist
```

Comment out the useless `COPY --from` and it works again.

**Comment by kolyshkin** [comment]
Filed https://github.com/moby/moby/issues/37965 as a simplified version of this one.

**Comment by jcberthon** [comment]
Wow that's being productive! :clap: 

I am fine with the simplified issue, so I will close this issue and monitor the other one.

Sadly, as I stated earlier, until mid-October my time is super limited, we have an important delivery and milestone achievement on one of my projects and I have to stay focus. I'll keep an eye on the new issue during that time but not on a regular basis. However after this delivery is past, I could dedicate several hours to this issue.

**Comment by GrantBirki** [comment]
I found an interesting fix for this! So if you run a COPY and the COPY doesn't actually change the state of the container it can fail. In my case the COPY may or may not actually change the container state. If it didn't it would fail. If it did it would work... 

So I had to ensure that the COPY always changed the state of the container. I did so like this in my case:

`# COPY Server Conf
RUN rm /etc/nginx/conf.d/nginx-server.conf
COPY nginx-server.conf /etc/nginx/conf.d/nginx-server.conf`

I just made sure to always remove the file I was COPYing first and that ensured the command always changed the state of the container.
